src/aurelian/agents/d4d_agent.py

```diff
@@ -0,0 +1,123 @@
+import asyncio
+import gradio as gr
+import requests
+import tempfile
+from pdfminer.high_level import extract_text
+from pydantic_ai import Agent, RunContext
+from aurelian.utils.search_utils import retrieve_web_page
+
+def get_full_schema(
+        url="https://raw.githubusercontent.com/monarch-initiative/ontogpt/main/src/ontogpt/templates/data_sheets_schema.yaml"
+    ) -> str:
+    """
+    Load the full datasheets for datasets schema from GitHub.
+    We use the raw URL so that we get plain text.
+    """
+    response = requests.get(url)
+    if response.status_code == 200:
+        return response.text
+    else:
+        return "Error: Unable to load full schema."
+
+
+FULL_SCHEMA = get_full_schema()
+
+# Create the agent using the full schema in the system prompt.
+data_sheets_agent = Agent(
+    model="openai:gpt-4o",
+    system_prompt=f"""
+Below is the complete datasheets for datasets schema:
+
+{FULL_SCHEMA}
+
+When provided with a URL to a webpage or PDF describing a dataset, your task is to fetch the 
+content, extract all the relevant metadata, and output a YAML document that exactly 
+conforms to the above schema. The output must be valid YAML with all required fields 
+filled in, following the schema exactly.
+""",
+)
+
+
+def safe_run(prompt: str):
+    """
+    Ensure an event loop is available and then call the agent's synchronous method.
+    """
+    try:
+        loop = asyncio.get_running_loop()
+    except RuntimeError:
+        loop = asyncio.new_event_loop()
+        asyncio.set_event_loop(loop)
+    return data_sheets_agent.run_sync(prompt)
+
+
+def extract_text_from_pdf(pdf_url: str) -> str:
+    """
+    Download and extract text from a PDF given its URL, using a temporary file.
+    """
+    response = requests.get(pdf_url)
+    if response.status_code != 200:
+        return "Error: Unable to retrieve PDF."
+
+    try:
+        with tempfile.NamedTemporaryFile(suffix=".pdf", delete=True) as temp_pdf:
+            temp_pdf.write(response.content)
+            temp_pdf.flush()  # Ensure all data is written before reading
+
+            text = extract_text(temp_pdf.name)
+            return text.strip() if text else "Error: No text extracted from PDF."
+
+    except Exception as e:
+        return f"Error extracting PDF text: {e}"
+
+
+
+def process_website_or_pdf(url: str) -> str:
```

Comment:
this could go in utils, we have some other methods for url->pdf->text there

Reply:
thanks, I moved this to `utils/pdf_fetcher.py` (or do you think it should go in `DOIFetcher`?)